src/main/java/com/puppycrawl/tools/checkstyle/site/XdocsTemplateParser.java

```diff
@@ -0,0 +1,202 @@
+// modify it under the terms of the GNU Lesser General Public
+// License as published by the Free Software Foundation; either
+// version 2.1 of the License, or (at your option) any later version.
+//
+// This library is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+// Lesser General Public License for more details.
+//
+// You should have received a copy of the GNU Lesser General Public
+// License along with this library; if not, write to the Free Software
+// Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA
+///////////////////////////////////////////////////////////////////////////////////////////////
+
+package com.puppycrawl.tools.checkstyle.site;
+
+import java.io.File;
+import java.io.IOException;
+import java.io.Reader;
+import java.io.StringReader;
+import java.io.StringWriter;
+import java.util.HashMap;
+import java.util.Locale;
+import java.util.Map;
+
+import javax.swing.text.html.HTML.Attribute;
+
+import org.apache.maven.doxia.macro.MacroExecutionException;
+import org.apache.maven.doxia.macro.MacroRequest;
+import org.apache.maven.doxia.macro.manager.MacroNotFoundException;
+import org.apache.maven.doxia.module.xdoc.XdocParser;
+import org.apache.maven.doxia.parser.ParseException;
+import org.apache.maven.doxia.parser.Parser;
+import org.apache.maven.doxia.sink.Sink;
+import org.codehaus.plexus.component.annotations.Component;
+import org.codehaus.plexus.util.IOUtil;
+import org.codehaus.plexus.util.xml.pull.XmlPullParser;
+
+import com.puppycrawl.tools.checkstyle.utils.CommonUtil;
+
+/**
+ * Parser for Checkstyle's xdoc templates.
+ * This parser is responsible for generating xdocs({@code .xml}) from the xdoc
+ * templates({@code .xml.template}). The templates are regular xdocs with custom
+ * macros for generating dynamic content - properties, examples, etc.
+ * This parser behaves just like the {@link XdocParser} with the difference that all
+ * elements apart from the {@code macro} element are copied as is to the output.
+ *
+ * @see ExampleMacro
+ */
+@Component(role = Parser.class, hint = "xdocs-template")
+public class XdocsTemplateParser extends XdocParser {
+
+    /** The macro parameters. */
+    private final Map<String, Object> macroParameters = new HashMap<>();
+
+    /** The source content of the input reader. Used to pass into macros. */
+    private String sourceContent;
+
+    /** A macro name. */
+    private String macroName;
+
+    @Override
+    public void parse(Reader source, Sink sink, String reference) throws ParseException {
+        try (Reader reader = source) {
+            final StringWriter contentWriter = new StringWriter();
+            IOUtil.copy(reader, contentWriter);
+            sourceContent = contentWriter.toString();
+            super.parse(new StringReader(sourceContent), sink, reference);
```

Comment:
We have reader as method argument, by writer we convert to strung.
Strung convert to reader again.
Can we just store reference to reader in class member? To convert it to string when we need

Reply:
`StringReader` doesn't override `toString`.

> [ERROR] Low: Method com.puppycrawl.tools.checkstyle.site.XdocsTemplateParser.processMacroEnd(Sink) calls toString() on an instance of a class that hasn't overridden toString() [com.puppycrawl.tools.checkstyle.site.XdocsTemplateParser] At XdocsTemplateParser.java:[line 180] SPP_NON_USEFUL_TOSTRING

From https://frontbackend.com/java/how-to-convert-reader-to-string-in-java
I tried using `com.google.common.io.CharStreams.toString()` and `org.apache.commons.io.IOUtils` but both fail.

Reply:
I ok to use special conversion to String when we need it.
Can we store reader as field of class and convert to string right before we need string content from it .

Reply:
I quickly hacked to user reader later, but looks like it closed or empty.

I surprised that even this is not working:
```
diff --git a/src/main/java/com/puppycrawl/tools/checkstyle/site/XdocsTemplateParser.java b/src/main/java/com/puppycrawl/tools/checkstyle/site/XdocsTemplateParser.java
index c31cfdd..835b041 100644
--- a/src/main/java/com/puppycrawl/tools/checkstyle/site/XdocsTemplateParser.java
+++ b/src/main/java/com/puppycrawl/tools/checkstyle/site/XdocsTemplateParser.java
@@ -22,7 +22,6 @@ package com.puppycrawl.tools.checkstyle.site;
 import java.io.File;
 import java.io.IOException;
 import java.io.Reader;
-import java.io.StringReader;
 import java.io.StringWriter;
 import java.util.HashMap;
 import java.util.Locale;
@@ -70,7 +69,7 @@ public class XdocsTemplateParser extends XdocParser {
         try (StringWriter contentWriter = new StringWriter()) {
             IOUtil.copy(source, contentWriter);
             sourceContent = contentWriter.toString();
-            super.parse(new StringReader(sourceContent), sink, reference);
+            super.parse(source, sink, reference);
         }
         catch (IOException ioException) {
             throw new ParseException("Error reading the input source", ioException);

```

Reply:
please double check , and I am to recall this review items.
As we use doxia we have to do crasy code to follow what they think is ok, hacking around their code.

Reply:
I can not get it any shorter/simpler here. I am out of ideas. @nrmancuso Can you please take a look?

Reply:
Resolved.